### PR TITLE
pkg/openthread: rework of FTD and MTD support

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This is a rework of #7149. The original PR by @biboc includes support for FTD, MTD and NCP.
This is (almost) and exact copy of his original work, but removes NCP support (see [discussion](https://github.com/RIOT-OS/RIOT/pull/7149#issuecomment-387392066)

It also sets the pkg version to the OpenThread Release [Thread Reference 2017-07-16](https://github.com/openthread/openthread/releases/tag/thread-reference-20170716).

I had to modify a couple of stuff in order to make it work with the Release version, and I pushed those commits as `fixup!`. The initial commit was already reviewed in #7149.

#7149 should still be open, because it contains the NCP implementation that can be used after this gets merged.

### Issues/PRs references
#7149

### CODE REVIEW

**Comment by jia200x** [comment]
Tested on IoTLAB and all multihop is working as expected. MTD is also working as it should.

File: Makefile.dep [hunk 1]
```diff
@@ -697,6 +697,11 @@ ifneq (,$(filter gcoap,$(USEMODULE)))
   USEMODULE += gnrc_sock_udp
 endif
 
+ifneq (,$(filter openthread,$(USEPKG)))
```
Comment (aabadie) [inline Makefile.dep]: I think the package dependencies could be added to a new `pkg/openthread/Makefile.dep` file instead. xtimer module could also be added as a dependency there.

File: tests/openthread/Makefile [hunk 2]
```diff
@@ -32,17 +47,12 @@ endif
 
 USEMODULE += $(DRIVER)
 
+USEMODULE += xtimer
```
Comment (aabadie) [inline tests/openthread/Makefile]: No need to add xtimer module as a dependency here if it's already a dependency of the openthread package.

**Comment by jia200x** [comment]
@aabadie done!

File: tests/openthread/Makefile [hunk 3]
```diff
@@ -12,11 +13,25 @@ RIOTBASE ?= $(CURDIR)/../..
...
+  # ftd: A Full Thread Device has router functionality compiled in
+  USEMODULE += openthread-ftd
+  USEMODULE += openthread-cli-ftd
+endif
+
+#Define PANID, CHANNEL and UART_BAUDRATE used by default
+OPENTHREAD_PANID ?= 0xbeef
+OPENTHREAD_CHANNEL ?= 26
+
+CFLAGS += -DOPENTHREAD_PANID=${OPENTHREAD_PANID}
```
Comment (aabadie) [inline tests/openthread/Makefile]: minor: use parenthesis instead of curly braces for accessing variables in Makefiles

File: tests/openthread/main.c [hunk 4]
```diff
@@ -10,25 +10,21 @@
...
 {
-    printf("Get PANID\n");
+    puts("This a test for OpenThread");
+    /* Example of how to call OpenThread stack functions */
+    puts("Get PANID ");
     uint16_t panid = 0;
     uint8_t res = ot_call_command("panid", NULL, (void*)&panid);
     printf("Current panid: 0x%x (res:%x)\n", panid, res);
 
-    openthread_uart_run();
```
Comment (aabadie) [inline tests/openthread/main.c]: why not keep the openthread shell ?

Comment (jia200x) [reply tests/openthread/main.c]: this is the OpenThread shell. The flow is `OpenThread Core <-> Openthread CLI <-> UART`

**Comment by jia200x** [comment]
@aabadie addressed

File: tests/openthread/main.c (hunk 4 above)
Comment (aabadie) [reply tests/openthread/main.c]: yes but here it's removed, so how can the OT shell be used ?

Comment (jia200x) [reply tests/openthread/main.c]: ah, I see what you mean. Now there's no "openthread_uart_run" simulator. It directly handles UART at interrupt level in the file platform_uart.c

Comment (aabadie) [reply tests/openthread/main.c]: Ok, got it now !

**Comment by jia200x** [comment]
@aabadie just addressed Murdock issues. Let's see how it goes

**Comment by aabadie** [comment]
@jia200x, seems like Murdock workers are missing automake.

**Comment by jia200x** [comment]
@aabadie I opened [this PR](https://github.com/RIOT-OS/riotdocker/pull/40) some time ago in the riotdocker repo. It adds automake

**Comment by Hyungsin** [comment]
Any update here?

**Comment by aabadie** [comment]
> I opened this PR some time ago in the riotdocker repo. It adds automake

I just merged it but now I think we have to wait for the new docker image to be deployed on Murdock workers.

**Comment by aabadie** [comment]
@jia200x, how about moving the openthread test application to the `examples` directory ?

**Comment by jia200x** [comment]
Back from holidays, sorry the delay.
It makes sense to me to move the test. I will do it.

**Comment by jia200x** [comment]
@aabadie just moved the examples folder.  Do you know if the Murdock workers were updated?

**Comment by aabadie** [comment]
`bootstrap: --automake: not found`: It's still the same problem, Murdock workers are not providing automake. We can't merge this PR because of this.

**Comment by jia200x** [comment]
Triggered Murdock again, the riotdocker image is updated. So the workers I guess

**Comment by jia200x** [comment]
@kaspar030 Murdock still doesn't find automake, although it's included as a dependency [here](https://github.com/RIOT-OS/riotdocker/blob/master/Dockerfile#L46) :(

Any clues?

**Comment by Hyungsin** [comment]
@jia200x, + conflict needs to be resolved.
@kaspar030, can you look at this murdock issue?

**Comment by miri64** [comment]
> @kaspar030 Murdock still doesn't find automake, although it's included as a dependency here :(

From what I can gather from the [script you are using](https://github.com/openthread/openthread/blob/master/third_party/nlbuild-autotools/repo/scripts/bootstrap) and murdock's output [LIBTOOLIZE](https://github.com/openthread/openthread/blob/master/third_party/nlbuild-autotools/repo/scripts/bootstrap#L159) is an empty string [here](https://github.com/openthread/openthread/blob/master/third_party/nlbuild-autotools/repo/scripts/bootstrap#L243).

**Comment by jia200x** [comment]
[I just opened a PR](https://github.com/RIOT-OS/riotdocker/pull/48) in the `riotdocker` repo that fixes the issue.

File: pkg/openthread/Makefile.include [hunk 5]
```diff
@@ -2,7 +2,9 @@ OPENTHREAD_DIR = $(RIOTBASE)/pkg/openthread
 
 INCLUDES += -I$(OPENTHREAD_DIR)/include \
             -I$(BINDIRBASE)/pkg/$(BOARD)/openthread/output/include \
+            -I$(BINDIRBASE)/pkg/$(BOARD)/openthread/include \
             -I$(BINDIRBASE)/pkg/$(BOARD)/openthread/include/openthread \
+            -I$(BINDIRBASE)/pkg/$(BOARD)/openthread/include/openthread/platform \
```
Comment (cladmi) [inline pkg/openthread/Makefile.include]: Why are this one and the previous one required? The one that are subdirectories from `include`.
It compiles without them and the C files use include with `openthread/header.h` and `openthread/platform/header.h`.

Also, when comparing the diff between the build directories with and without the lines, the generated `.d` files are the same so suggest that indeed it did not change any include path.

Comment (cladmi) [reply pkg/openthread/Makefile.include]: Also, for checking right now, I think there is not reason to depend on `openthread/output/include` directory.
(except that the `sed` on `types.h` should be done with a patch in the main `include` directory).

It would also be better because having other modules depending on openthread being compiled before them to have the `output` directory is not something entirely guaranteed. Normally only the `prepare` has been done before.
For packages there is still a rule to build them before other modules but it is more a side-effect than a real ensured property. Also another package that would depend on this header would have problem as it may not exist.

**Comment by jia200x** [comment]
@cladmi the changes were addressed

I also rebased and fixed the `uart_stdio.h` to `stdio_uart.h` rename.

**Comment by aabadie** [comment]
@jia200x, there's a problem with your rebase: >1k files changed. All tests application have been moved to examples directory. Weird.

**Comment by aabadie** [comment]
in fact, the problem is 90adb9ef60aee96db1224d75b4458f5f5fa83a2a. You git mv all tests applications to examples...

**Comment by jia200x** [comment]
> You git mv all tests applications to examples...

I did a rename + fixup and this is the result. I'm currently fixing it

**Comment by jia200x** [comment]
@aabadie I discovered what happened. The tests folder was refactored, and when doing rebase the `move test to examples` commit did this mess. Nice way to spend an afternoon evening :)

It should be ok now

**Comment by jia200x** [comment]
@aabadie @cladmi could you do a final review/test so we are one step ahead to merge? :)

File: examples/openthread/0001-fixup-pkg-openthread-add-FTD-and-NCP-support.patch [hunk 6]
```diff
@@ -0,0 +1,24 @@
+From 4dba7adf797e49350c750c48acd3ed47eecc19e5 Mon Sep 17 00:00:00 2001
```
Comment (cladmi) [inline examples/openthread/0001-fixup-pkg-openthread-add-FTD-and-NCP-support.patch]: These file should not be there.

File: examples/openthread/0001-pkg-openthread-adapt-to-newer-stdio-uart-header-file.patch [hunk 7]
```diff
@@ -0,0 +1,50 @@
+From ddcdfe5be66f62b032a08adfcc207661fe2735d2 Mon Sep 17 00:00:00 2001
```
Comment (cladmi) [inline examples/openthread/0001-pkg-openthread-adapt-to-newer-stdio-uart-header-file.patch]: This one should not be there too.

File: examples/openthread/0001-pkg-openthread-replace-sed-by-patch.patch [hunk 8]
```diff
@@ -0,0 +1,44 @@
+From 65d3ea906c4908cc8382f0f4dad8e89697918908 Mon Sep 17 00:00:00 2001
```
Comment (cladmi) [inline examples/openthread/0001-pkg-openthread-replace-sed-by-patch.patch]: And this patch neither.

File: tests/openthread/Makefile (hunk 2 above)
Comment (cladmi) [reply tests/openthread/Makefile]: Still there.

File: examples/openthread/0001-fixup-pkg-openthread-add-FTD-and-NCP-support.patch (hunk 6 above)
Comment (jia200x) [reply examples/openthread/0001-fixup-pkg-openthread-add-FTD-and-NCP-support.patch]: my mistake, I push them by error

**Comment by jia200x** [comment]
@cladmi done

**Comment by jia200x** [comment]
> @jia200x Please do not rebase / amend during a review, it makes it hard to follow changes. I basically need to restart a review from scratch everytime.

Ok, sorry

File: examples/openthread/README.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
+## OpenThread on RIOT
+
+This test demonstrates how to use the [OpenThread](https://github.com/openthread/openthread)
```
Comment (aabadie) [inline examples/openthread/README.md]: test ?

File: examples/openthread/README.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+open source implementation of [Thread](https://threadgroup.org/) on RIOT.
+ The [Command Line Interface](https://github.com/openthread/openthread/blob/master/examples/apps/cli/README.md) of
```
Comment (aabadie) [inline examples/openthread/README.md]: Minor: there is an extra space at the beginning of the line

File: examples/openthread/README.md (hunk 9 above)
Comment (jia200x) [reply examples/openthread/README.md]: thanks, forgot this one.

File: examples/openthread/README.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+(https://github.com/openthread/openthread/blob/master/src/cli/README.md) of the CLI for usage information.
+
+You can either build a FTD or MTD firmware:
+- MTD: A Minimal Thread Device does not have router functionality compiled in. An MTD may or may not be sleepy.
+- FTD: A Full Thread Device has router functionality compiled in.
+
+## Quick usage
+
+With RIOT port, a node is auto-setup and ready to communicate with this configuration:
+OPENTHREAD_PANID: 0xbeef
```
Comment (aabadie) [inline examples/openthread/README.md]: Maybe surround the configuration with triple back ticks. I would also write them using the makefile variable assignment syntax, e.g with `=`

File: examples/openthread/README.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+OPENTHREAD_CHANNEL: 26
+
+You can pass the panid/channel independently when building the firmware:
+`make BOARD=<target> OPENTHREAD_PANID=0xaaaa OPENTHREAD_TYPE=ftd flash term`
```
Comment (aabadie) [inline examples/openthread/README.md]: Same here, surround the commands in triple back ticks

File: examples/openthread/README.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+`make BOARD=<target> OPENTHREAD_CHANNEL=20 OPENTHREAD_TYPE=ftd flash term`
+
+To test OpenThread in RIOT, you can do the following:
+
+1. Flash nodes with MTD or FTD functionality:
+`make BOARD=<target> clean all flash OPENTHREAD_TYPE=mtd`
```
Comment (aabadie) [inline examples/openthread/README.md]: same here with the triple back quotes

File: examples/openthread/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+   fe80::3984:f4eb:d182:5dae
+```
+5. Ping from another node with `ping fdde:ad00:beef:0:946a:c722:a5d9:848`.
+6. You can try IEEE802.15.4 scan with `scan` command
+7. You can also check other commands with `help`
+
+
+## OpenThread port on RIOT status
+
+OpenThread port on RIOT is stable. In case of any bug, please report via GitHub issue.
```
Comment (aabadie) [inline examples/openthread/README.md]: You can use a link for `GitHub issue`, like `https://github.com/RIOT-OS/RIOT/issues/new?template=bug_report.md&title=Bug:`

**Comment by jia200x** [comment]
@aabadie fixed README.md. I also change it to make it a little bit more compliant to the 80 chars limit (although the first links make it imposible to reach 80 chars always)

**Comment by jia200x** [comment]
(also added a `\n` after every triple backquotes, otherwise is not detected as block)

File: examples/openthread/README.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
+## OpenThread on RIOT
+
+This example demonstrates how to use the [OpenThread](https://github.com/openthread/openthread)
+open source implementation of [Thread](https://threadgroup.org/) on RIOT.
+The [Command Line Interface](https://github.com/openthread/openthread/blob/master/examples/apps/cli/README.md) of
+OpenThread was ported. Please check the [full documentation]
+(https://github.com/openthread/openthread/blob/master/src/cli/README.md)
```
Comment (aabadie) [inline examples/openthread/README.md]: The link to full documentation is broken, see [here](https://github.com/jia200x/RIOT/tree/openthread_ftd/examples/openthread#openthread-on-riot)
If you want to separate the paragraph, a new line is not enough, you have to add a blank line.

Comment (jia200x) [reply examples/openthread/README.md]: right. Just pushed the fix (https://github.com/RIOT-OS/RIOT/pull/9336/commits/fd96a55f9ab4a9d5c220910931b94fc3e41f97fc)

**Comment by jia200x** [comment]
@smling @kaspar030 for some reason Murdock still complains about automake, m4, libtool stuff :(
I tried to compile it with BUILD_IN_DOCKER=1 and the image works.

Any clue?

**Comment by aabadie** [comment]
> Any clue?

The Murdock workers need to be updated with the latest docker image version. This is not done automatically and other remaining things may change apparently (use ubuntu bionic, etc)

**Comment by jia200x** [comment]
> The Murdock workers need to be updated with the latest docker image version. This is not done automatically and other remaining things may change apparently (use ubuntu bionic, etc)

ping? @kaspar030 @smlng

**Comment by jia200x** [comment]
awesome! I will squash then ;)

**Comment by jia200x** [comment]
@aabadie there's an unrelated issue with pkg_libhydrogen

**Comment by aabadie** [comment]
> there's an unrelated issue with pkg_libhydrogen

Saw that. I'm wondering why execution tests are run when only `CI:ready for build` label is set. Any idea ?

**Comment by Hyungsin** [comment]
This is very encouraging! Thank you guys! 👍

**Comment by biboc** [comment]
Good job @jia200x ! 

How about NCP port? 
I do think it is important since it can replace the actual RIOT BorderRouter and the port will be almost completed
Last thing missing was commissioning if I remember and I had problems with amount of free RAM for samr21-xpro
